### DISCUSSION TITLE
The Hold-off Trigger Control

### DISCUSSION THREAD

**Discussion by KanKub** [OpenHantek/OpenHantek6022#197] [Q&A] [answered]
Is it possible to implement hold-off control to 6022BE? it would be a lot easier to diagnose the logic signal.

**Comment by Ho-Ro** [comment, marked as answer]
As the Hantek has no HW trigger everything has to be done in SW - fetch one block, search for a trigger condition and display. So no chance to delay more than one screen width. Sorry, this is a very limited HW.
But if you want to check logic a [simple LA](https://sigrok.org/wiki/Lcsoft_Mini_Board) would be more sufficient.

**Reply by bobatsar** [reply]
@Ho-Ro I would also like to have a hold-off trigger as I use this quite often and I don't understand what's the problem here.

Shouldn't it be possible to remember the time when the last trigger triggered, then disable the trigger detection for the hold-off time. When the hold-off time is elapsed, start the next trigger.

Just like in the first example on https://rfmw.em.keysight.com/wireless/helpfiles/89600b/webhelp/subsystems/gui/Content/input_trigger_holdoff.htm

**Reply by Ho-Ro** [reply]
Did you check this setting?
![image](https://user-images.githubusercontent.com/12542359/146531861-52634e7b-72d7-4d19-9e8f-ca1b90b701f0.png)
Maybe it helps.

> Shouldn't it be possible ...

Of course - please feel free to experiment. I am always happy to receive input, preferably as a pull request. As a starting point you could check [this documentatation](https://github.com/OpenHantek/OpenHantek6022/blob/main/docs/developer_info.md#data-flow).

The magic happens in `hantekdsocontrol.cpp` in
[`unsigned HantekDsoControl::searchTriggeredPosition()`](https://github.com/OpenHantek/OpenHantek6022/blob/31c5ee2d80fd62366c0e0818ef275809be8d5f58/openhantek/src/hantekdso/hantekdsocontrol.cpp#L650)
and 
[`bool HantekDsoControl::provideTriggeredData()`](https://github.com/OpenHantek/OpenHantek6022/blob/31c5ee2d80fd62366c0e0818ef275809be8d5f58/openhantek/src/hantekdso/hantekdsocontrol.cpp#L705)

Use `--verbose 5` or `--verbose 6` to follow the trigger detection (use slow signals, it gives **a lot** of output), see snippet below.
````
...
    HDC::stateMachine() 8
    HDC::convertRawDataToSamples() 8
    HDC::searchTriggeredPosition() 8
     HDC::searchTriggerPoint() 0 0
     pre: 5000 post: 15000
     HDC::searchTriggerPoint() 1 6182
     pre: 6182 post: 16182
     HDC::searchTriggerPoint() 0 6964
     pre: 6964 post: 16964
     nextSlope: / triggeredPositionRaw: 6182
    HDC::provideTriggeredData() 8
    HDC::getRecordLength() -> 2002944
    PostProcessing::input() 8
    SpectrumGenerator::process() 8
     SpectrumGenerator::process() 0 freq: 13 32.5 corr: 1562 32.0102
     SpectrumGenerator::process() 1 freq: 14 35 corr: 145 344.828
    GraphGenerator::process() 8
     GraphGenerator::generateGraphsTYvoltage() 8
     GraphGenerator::generateGraphsTYspectrum() 8
     MainWindow::showNewData() 8
    DsoWidget::showNew() 8
    HDC::stateMachine() 8
...
````
With your proof of concept I will implement it into the gui.

**Reply by bobatsar** [reply]
Ok thanks for the hints.
I will check it and see if I can find a solution.
